### PR TITLE
Saveable Generator: initial import

### PR DESCRIPTION
**Pull request by flexthink**
## What does this PR do?

This PR improves reproducibility for scenarios where training or inference can be highly non-deterministic, especially in cases where the selection of data, layers to train, etc is done randomly and it is unreasonable to expect training to complete in a single run.

It provides an opt-in wrapper called `SaveableGenerator` that will save the state of the available global generations (or custom generators if specified) together with the checkpoint.

Sample usage:

```yaml
generator: !new:model.custom_model.SaveableGenerator # <-- Include the wrapper

checkpointer: !new:speechbrain.utils.checkpoints.Checkpointer
    checkpoints_dir: !ref <save_folder>
    recoverables:
        model: !ref <model>
        lr_scheduler: !ref <lr_annealing>
        counter: !ref <epoch_counter>
        generator: !ref <generator> # <-- Add to the checkpoint
```

Usage is entirely opt-in. There are no changes to existing recipes or libraries.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/utils/repro.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    consider implementing a custom generator.
+
+    Running it on an unsupported device not using the Torch
+    generator interface will simply fail to restore the
+    state but will not cause an error.
+
+    Arguments
+    ---------
+    generators : list, optional
+        A list of generator objects. If not provided,
```
Comment (pplantinga) [inline speechbrain/utils/repro.py]: Sentence not complete, what happens if not provided?

Comment (pplantinga) [reply speechbrain/utils/repro.py]: Also, can you add an example to the docstring?

Comment (pplantinga) [reply speechbrain/utils/repro.py]: Also, I think the type should be `dict` or `Mapping` instead of `list`, and this should describe how the mapping should be defined, e.g. `Mapping[str, Generator]` or something like this. I think the torch generator is of the generic type:

https://docs.python.org/3/library/collections.abc.html#collections.abc.Generator

File: tests/unittests/test_repro.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+"""Unit tests for reproducibility utilities"""
+
+import torch
+
+
+def test_repro(tmpdir):
```
Comment (pplantinga) [inline tests/unittests/test_repro.py:8]: Some of our unit tests take a device parameter for testing on e.g. cuda. Perhaps we can do something similar here to ensure it works at least locally (I guess the CI is running on cpu).

**Review by pplantinga — COMMENTED** [review]
I like the idea of saving the random generator state and this seems like a nice implementation to me. Am I understanding correctly that one could fix a recipe to   restart the random seed correctly just by adding a `SaveableGenerator` object to the checkpointer without any arguments? Maybe we could consider a separate PR to convert many of our recipes.

File: tests/unittests/test_repro.py (hunk 2 above)
Comment (flexthink) [reply tests/unittests/test_repro.py:8]: @pplantinga : Done in a separate test. However, device support is currently limited given that:
- Torch has inconsistencies in how generators are handled
- Even for Cuda, default generators are no longer exposed *as generators*
- Some devices like MPS don't even expose RNG state

So for now, this feature will be only for the most common use cases. Other devices can be added later by writing wrappers similar to the one I had for Cuda - if they support the functionality at all.

File: speechbrain/utils/repro.py (hunk 1 above)
Comment (flexthink) [reply speechbrain/utils/repro.py]: @pplantinga : Addressed... However, in reality, this takes a "generator-like object", only getting/setting the state is required - no need to implement the full interface.

**Comment by flexthink** [comment]
> I like the idea of saving the random generator state and this seems like a nice implementation to me. Am I understanding correctly that one could fix a recipe to restart the random seed correctly just by adding a `SaveableGenerator` object to the checkpointer without any arguments? Maybe we could consider a separate PR to convert many of our recipes.

That is absolutely correct. Further customization may be needed if the recipes use custom generators, but for the base case that's all it takes.

**Review by pplantinga — APPROVED** [review]
This LGTM!
